test/Error/Labels.js

```diff
@@ -0,0 +1,81 @@
+    `a: 
+        function x()
+        {
+        }`
+testVarAfterLabel = 
+    `b:
+        var a;`
+testForLoopAfterLabel = 
+    `c:
+        for(i = 0; i < 10; i++)
+        {
+            d:
+                for(j = 0; j < 10; j++)
+                {
+                    if((i%j)%3 === 0)
+                        break d;
+                }
+        }`
+testWhileLoopAfterLabel = 
+    `i = 1, j = 1, product = 1
+    c:
+        while(i <= 20)
+        {
+            d:
+                while(j <= 5)
+                {
+                    j = j+2
+                    product *= i*j;
+                    break c;
+                }
+            i = i+1
+        }
+        assert.areEqual(3, product)
+        `
+testNonLexicalDeclarationAfterLabel=
+    `d:
+        g = "Hola, Mundo."`
+var tests = [
+    {
+        name : "Labels followed by declarations which should all be syntax errors.",
+        body : function () 
+        {
+            assert.throws(() => eval("a: let x;"), SyntaxError, 
+            "A let declaration must not be labelled", "Labels not allowed before lexical declaration");
+            assert.throws(() => eval("b: const y = 0;"), SyntaxError, 
+            "A const declaration must not be labelled", "Labels not allowed before lexical declaration");
+            assert.throws(() => eval("c: class z {}"), SyntaxError, 
+            "An class declaration must not be labelled", "Labels not allowed before class declaration");
+            assert.throws(() => eval("d: function* w() {}"), SyntaxError, 
+            "A generator must not be labelled", "Labels not allowed before generator declaration");
+            assert.throws(() => eval("e: async function u() { }"), SyntaxError, 
+            "An async function must not be labelled", "Labels not allowed before async function declaration");
+            /*assert.throws(() => eval("e: async function* u() { }"), SyntaxError, //TODO: Uncomment when async generators supported.
+            "An async generator must not be labelled", "Labels not allowed before async function declaration");*/
+        }
+    },
+    {
+        name : "Labels followed by non-declarations are valid syntax (non-generator non-async function declarations and non-lexical variable declarations are allowed)",
+        body : function () 
+        {
+            assert.doesNotThrow(() => eval(testFunctionAfterLabel), "Function declarations should be allowed after a label."); 
+            assert.doesNotThrow(() => eval(testVarAfterLabel), "var declarations should be allowed after a label.");
+            assert.doesNotThrow(() => eval(testForLoopAfterLabel), "for loop statements should be allowed after a label.");
+            assert.doesNotThrow(() => eval(testNonLexicalDeclarationAfterLabel), "non-lexical declarations should be allowed after a label.");
+            assert.doesNotThrow(() => eval(testWhileLoopAfterLabel), "while loops should be allowed after a label");
+        }
+    }
+];

```

Comment:
Would you mind adding tests for some of the strict mode specific stuff? E.g.

https://tc39.github.io/ecma262/#sec-labelled-function-declarations
Note
The early error rules for WithStatement, IfStatement, and IterationStatement prevent these statements from containing a labelled FunctionDeclaration in non-strict code.

Reply:
Some other interesting cases are in https://tc39.github.io/ecma262/#sec-identifiers-static-semantics-early-errors, e.g. yield and async as label names depending on context and/or strict mode.

Reply:
Added test cases for strict mode